File: BAEKJOON/Javascript/1000/1900/1926.js
/*
Title : 그림
Link : https://www.acmicpc.net/problem/1926
*/

const fs = require("fs");
const filepath =
  process.platform === "linux" ? "/dev/stdin" : __dirname + "/input.txt";
const input = fs.readFileSync(filepath).toString().trim().split("\n");

class BFS {
  constructor(board, n, m) {
    this.board = board;
    this.n = n;
    this.m = m;
    this.visited = Array.from({ length: n + 2 }, () =>
      Array(m + 2).fill(false)
    );
    this.delta = [
      [0, 1],
      [0, -1],
      [1, 0],
      [-1, 0],
    ];
    this.pieces = 0;
    this.maxSize = 0;
  }

  search() {
    for (let i = 1; i <= this.n; i++) {
      for (let j = 1; j <= this.m; j++) {
        if (this.board[i][j] === 1 && !this.visited[i][j]) {
          this.bfs(i, j);
          this.pieces++;
        }
      }
    }
    console.log(this.pieces);
    console.log(this.maxSize);
  }

  bfs(i, j) {
    let size = 0;
    this.visited[i][j] = true;
    let queue = [[i, j]];

    while (queue.length) {
      const [x, y] = queue.pop();
      size++;

      for (let i = 0; i < 4; i++) {
        const nx = x + this.delta[i][0];
        const ny = y + this.delta[i][1];

        if (this.board[nx][ny] === 1 && !this.visited[nx][ny]) {
          this.visited[nx][ny] = true;
          queue.push([nx, ny]);
        }
      }
    }
    this.maxSize = Math.max(this.maxSize, size);
  }
}

const [n, m] = input[0].split(" ").map(Number);
let board = Array.from({ length: n + 2 }, () => Array(m + 2).fill(0));
for (let i = 1; i <= n; i++) {
  board[i] = [0].concat(input[i].split(" ").map(Number)).concat([0]);
}

const bfs = new BFS(board, n, m);
bfs.search();

File: BAEKJOON/Python/1000/1900/1926.py
"""
Title : 그림
Link : https://www.acmicpc.net/problem/1926
"""

from sys import stdin

input = stdin.readline


class BFS:
    def __init__(self, board: list[list[int]], n: int, m: int) -> None:
        self.board = board
        self.n = n
        self.m = m
        self.visited = [[False for _ in range(m + 2)] for _ in range(n + 2)]
        self.delta = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        self.pieces = 0
        self.max_size = 0

    def search(self) -> None:
        for i in range(1, len(self.board) - 1):
            for j in range(1, len(self.board[0]) - 1):
                if self.board[i][j] == 1 and not self.visited[i][j]:
                    self.pieces += 1
                    self.max_size = max(self.max_size, self.bfs(i, j))

    def bfs(self, x: int, y: int) -> int:
        size = 1
        self.visited[x][y] = True
        queue = [(x, y)]
        while queue:
            x, y = queue.pop(0)
            for dx, dy in self.delta:
                nx, ny = x + dx, y + dy
                if self.board[nx][ny] == 1 and not self.visited[nx][ny]:
                    self.visited[nx][ny] = True
                    size += 1
                    queue.append((nx, ny))
        return size


if __name__ == "__main__":
    n, m = map(int, input().split())
    board = [list(map(int, input().split())) for _ in range(n)]
    board = [[0] * (m + 2)] + [[0] + i + [0] for i in board] + [[0] * (m + 2)]

    bfs = BFS(board, n, m)
    bfs.search()
    print(f"{bfs.pieces }\n{bfs.max_size}")
